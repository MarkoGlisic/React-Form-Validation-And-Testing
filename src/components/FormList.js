import { useRef } from "react";
import {
  Form,
  Container,
  InputGroup,
  FormControl,
  Button,
} from "react-bootstrap";

const FormList = () => {
  const nameRef = useRef();
  const emailRef = useRef();
  const ageRef = useRef();

  const nameValidation = (name) => {
    if (/[^a-zA-Z -]/.test(name)) return console.log("Invalid Character");
    if (name.trim().length < 3)
      return console.log("Name needs to be at least three characters long");
    else return true;
  };

  const emailValidation = (email) => {
    if (
      !/^[a-zA-Z0-9.!#$%&â€™*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(
        email
      )
    ) {
      return console.log("Invalid E-mail Format");
    } else return true;
  };

  const ageValidation = (age) => {
    if (!/^[0-9]*$/.test(age))
      return console.log("Age must be a numeric value");
    if (age < 18) return console.log("Age must be at least 18");
    if (age > 99) return console.log("Age must be less than 99");
    else return true;
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (
      nameValidation(nameRef.current.value) &&
      emailValidation(emailRef.current.value) &&
      ageValidation(ageRef.current.value)
    )
      return console.log("Succesfully Submitted");
  };
  return (
    <>
      <h3>Please enter the values:</h3>
      <Container className="mt-3">
        <Form data-testid="form" onSubmit={handleSubmit}>
          <InputGroup className="mb-3">
            <InputGroup.Prepend>
              <InputGroup.Text>Enter your name</InputGroup.Text>
            </InputGroup.Prepend>
            <FormControl
              data-testid="name-input"
              type="text"
              ref={nameRef}
              required
            />
          </InputGroup>
          <InputGroup className="mb-3">
            <InputGroup.Prepend>
              <InputGroup.Text>Enter your email</InputGroup.Text>
            </InputGroup.Prepend>
            <FormControl
              data-testid="email-input"
              type="email"
              ref={emailRef}
              required
            />
          </InputGroup>
          <InputGroup className="mb-3">
            <InputGroup.Prepend>
              <InputGroup.Text>Enter your age</InputGroup.Text>
            </InputGroup.Prepend>
            <FormControl
              data-testid="age-input"
              type="text"
              ref={ageRef}
              required
            />
          </InputGroup>
          <Button
            data-testid="button-test"
            size="lg"
            variant="primary"
            type="submit"
          >
            Submit
          </Button>
        </Form>
      </Container>
    </>
  );
};

export default FormList;
