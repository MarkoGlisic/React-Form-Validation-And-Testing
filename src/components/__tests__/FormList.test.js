import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import FormList from "../FormList.js";
import renderer from "react-test-renderer";

afterEach(cleanup);

describe("<FormList/>", () => {
  // Email Input Tests
  it("renders email input", () => {
    render(<FormList />);
    const inputEl = screen.getByTestId("email-input");
    expect(inputEl).toBeInTheDocument();
    expect(inputEl).toHaveAttribute("type", "email");
  });
  it("pass a valid email to email input field", () => {
    render(<FormList />);
    const inputEl = screen.getByTestId("email-input");
    userEvent.type(inputEl, "test@test.com");
    expect(screen.getByTestId("email-input")).toHaveValue("test@test.com");
  });

  // Name Input Tests
  it("renders name input", () => {
    render(<FormList />);
    const inputEl = screen.getByTestId("name-input");
    expect(inputEl).toBeInTheDocument();
    expect(inputEl).toHaveAttribute("type", "text");
  });
  it("pass a valid name to name input field", () => {
    render(<FormList />);
    const inputEl = screen.getByTestId("name-input");
    userEvent.type(inputEl, "John");
    expect(screen.getByTestId("name-input")).toHaveValue("John");
  });

  // Age Input Tests
  it("renders name input", () => {
    render(<FormList />);
    const inputEl = screen.getByTestId("age-input");
    expect(inputEl).toBeInTheDocument();
    expect(inputEl).toHaveAttribute("type", "text");
  });
  it("pass a valid age to age input field", () => {
    render(<FormList />);
    const inputEl = screen.getByTestId("age-input");
    userEvent.type(inputEl, "25");
    expect(screen.getByTestId("age-input")).toHaveValue("25");
  });

  // Snapshots
  it("matches the snapshot", () => {
    const tree = renderer.create(<FormList />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
