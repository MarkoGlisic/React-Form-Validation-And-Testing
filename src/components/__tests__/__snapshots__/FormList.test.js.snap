// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormList/> matches the snapshot 1`] = `
Array [
  <h3>
    Please enter the values:
  </h3>,
  <div
    className="mt-3 container"
  >
    <form
      className=""
      data-testid="form"
      onSubmit={[Function]}
    >
      <div
        className="mb-3 input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
            Enter your name
          </span>
        </div>
        <input
          className="form-control"
          data-testid="name-input"
          required={true}
          type="text"
        />
      </div>
      <div
        className="mb-3 input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
            Enter your email
          </span>
        </div>
        <input
          className="form-control"
          data-testid="email-input"
          required={true}
          type="email"
        />
      </div>
      <div
        className="mb-3 input-group"
      >
        <div
          className="input-group-prepend"
        >
          <span
            className="input-group-text"
          >
            Enter your age
          </span>
        </div>
        <input
          className="form-control"
          data-testid="age-input"
          required={true}
          type="text"
        />
      </div>
      <button
        className="btn btn-primary btn-lg"
        data-testid="button-test"
        disabled={false}
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>,
]
`;
